### PR TITLE
Add update-vendor.sh

### PR DESCRIPTION
**Pull request by losipiuk**
A script for automatic vendor directory update in world when
k8s.io/kubernets uses go modules.

### CODE REVIEW

File: cluster-autoscaler/hack/update-vendor.sh [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+echo "Sending logs to: ${LOG_FILE}"
+if [ -z "${BASH_XTRACEFD:-}" ]; then
+  exec 19> "${LOG_FILE}"
+  export BASH_XTRACEFD="19"
+  set -x
+fi
+
+EXPECTED_ERROR_MARKER="${WORK_DIR}/expected_error"
+
+# Try
```
Comment (mwielgus) [inline cluster-autoscaler/hack/update-vendor.sh]: Try what?

Comment (jkaniuk) [reply cluster-autoscaler/hack/update-vendor.sh]: Try indented block starting with 
set +o errexit
(
  set -o errexit

fail on first error, then enter Catch block below if error occured

File: cluster-autoscaler/hack/update-vendor.sh [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+  git checkout ${K8S_REV} >>${LOG_FILE} 2>&1
+  K8S_REV_PARSED=$(git rev-parse ${K8S_REV})
+  popd >/dev/null
+
+  if [[ $(basename $(pwd)) != "cluster-autoscaler" ]];then
+    echo "The script must be run in cluster-autoscaler directory"
+    exit 1
+  fi
+
+  function err_rerun() {
```
[the remark this answers is not in the document]
Comment (losipiuk) [reply cluster-autoscaler/hack/update-vendor.sh]: Well - what it does - it prints error message and exits :)

[the remark this answers is not in the document]
Comment (jkaniuk) [reply cluster-autoscaler/hack/update-vendor.sh]: then it is an overkill

[the remark this answers is not in the document]
Comment (jkaniuk) [reply cluster-autoscaler/hack/update-vendor.sh]: if you need oneliner:
echo "error msg!"; exit 1

[the remark this answers is not in the document]
Comment (losipiuk) [reply cluster-autoscaler/hack/update-vendor.sh]: Not really - we still need to distinguish from expected and unexpected error. 
And with that being true I prefer to keep all the handling of one and the other in single place (expected in err_rerun and unexpected in catch clause). IMO it is easier to read that way.

File: cluster-autoscaler/hack/update-vendor.sh [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+  set -o errexit
+  rm -f $EXPECTED_ERROR_MARKER
+  K8S_REPO="${WORK_DIR}/kubernetes"
+  if [ -d ${K8S_REPO} ]; then
+    pushd ${K8S_REPO} >/dev/null
+    if [[ "$(git remote get-url origin)" != "${K8S_FORK}" ]]; then
+      echo "Mismated checked out k8s repo; deleting"
+      rm -rf "${K8S_REPO}"
+    fi
+    popd >/dev/null
```
[the remark this answers is not in the document]
Comment (losipiuk) [reply cluster-autoscaler/hack/update-vendor.sh]: You could use subshell here. But below (when K8S_REV_PARSED is modified) you could. Let's keep pushd popd consistently. I do not think parens whould change much in terms of readability here. It is bash anyway - so it is write only ;)

File: cluster-autoscaler/hack/update-vendor.sh [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    echo "Cloning ${K8S_FORK} into ${K8S_REPO}"
+    git clone ${K8S_FORK} ${K8S_REPO} >>${LOG_FILE} 2>&1
+  fi
+
+  pushd ${K8S_REPO} >/dev/null
+  git checkout ${K8S_REV} >>${LOG_FILE} 2>&1
+  K8S_REV_PARSED=$(git rev-parse ${K8S_REV})
+  popd >/dev/null
+
+  if [[ $(basename $(pwd)) != "cluster-autoscaler" ]];then
```
[the remark this answers is not in the document]
Comment (losipiuk) [reply cluster-autoscaler/hack/update-vendor.sh]: Yeah - moved

File: cluster-autoscaler/hack/update-vendor.sh [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+  # Fix module name and staging modules links
+  sed -i "s#module k8s.io/kubernetes#module k8s.io/autoscaler/cluster-autoscaler#" go.mod
+  sed -i "s#\\./staging#${K8S_REPO}/staging#" go.mod
+
+  function list_dependencies() {
+    local_tmp_dir=$(mktemp -d "${WORK_DIR}/list_dependencies.XXXX")
+    local go_dep_file="$1"
+    local tmp_file="${local_tmp_dir}/list_dependencies.tmp"
+    rm -f ${tmp_file}
+    go mod edit -json ${go_dep_file} |jq -r '.Replace[]? | select(.New.Version != null)| "\(.Old.Path) \(.New.Version)"' >> ${tmp_file}
```
[the remark this answers is not in the document]
Comment (losipiuk) [reply cluster-autoscaler/hack/update-vendor.sh]: Done

**Pull request merged**